File: app/src/main/java/com/ytempest/wanandroid/activity/register/RegisterActivity.java
package com.ytempest.wanandroid.activity.register;

import android.content.Context;
import android.os.Bundle;
import androidx.annotation.Nullable;
import android.text.InputFilter;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import com.ytempest.layoutinjector.annotation.InjectLayout;
import com.ytempest.tool.util.RegexUtils;
import com.ytempest.wanandroid.R;
import com.ytempest.wanandroid.base.activity.MvpActivity;
import com.ytempest.wanandroid.http.ErrCode;
import com.ytempest.wanandroid.http.bean.LoginBean;
import com.ytempest.wanandroid.listener.TextWatcherListener;
import com.ytempest.wanandroid.utils.SpaceInputFilter;
import com.ytempest.wanandroid.utils.StatusBarUtil;
import com.ytempest.wanandroid.widget.ModifiableButton;

import butterknife.BindView;
import butterknife.OnClick;

/**
 * @author heqidu
 * @since 2020/8/14
 */
@InjectLayout(R.layout.activity_register)
public class RegisterActivity extends MvpActivity<RegisterPresenter> implements IRegisterContract.View {

    @BindView(R.id.ll_register_root)
    ViewGroup mRootView;
    @BindView(R.id.et_register_account)
    EditText mAccountET;
    @BindView(R.id.et_register_pwd)
    EditText mPwdET;
    @BindView(R.id.et_register_pwd_confirm)
    EditText mPwdConfirmET;
    @BindView(R.id.view_register)
    ModifiableButton mRegisterButton;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        StatusBarUtil.immersive(this);
        StatusBarUtil.setPaddingSmart(this,mRootView);
        InputFilter[] filters = new InputFilter[]{new SpaceInputFilter()};
        mAccountET.setFilters(filters);
        mPwdET.setFilters(filters);
        mPwdConfirmET.setFilters(filters);

        TextWatcherListener textWatcher = new TextWatcherListener() {
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                super.onTextChanged(s, start, before, count);
                if (mAccountET.length() > 0 && mPwdET.length() > 0 && mPwdConfirmET.length() > 0) {
                    mRegisterButton.setNormalStatus();
                } else {
                    mRegisterButton.setDisableStatus();
                }
            }
        };
        mAccountET.addTextChangedListener(textWatcher);
        mPwdET.addTextChangedListener(textWatcher);
        mPwdConfirmET.addTextChangedListener(textWatcher);

        showSoftKeyBoard(mAccountET);
    }

    /**
     * 自动对焦到输入框，并打开软键盘
     */
    private void showSoftKeyBoard(EditText editText) {
        // 延迟200毫秒弹出，以防止自动弹出失效
        editText.postDelayed(() -> {
            InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            if (inputMethodManager != null) {
                editText.requestFocus();
                inputMethodManager.showSoftInput(editText, 0);
            }
        }, 200);
    }

    @OnClick(R.id.iv_register_back)
    void onBackClick() {
        finish();
    }

    @OnClick(R.id.view_register)
    void onRegisterClick() {
        int accountMinLen = getResources().getInteger(R.integer.account_min_len);
        String account = mAccountET.getText().toString().trim();
        if (account.length() < accountMinLen) {
            showToast(getString(R.string.account_len_limit, accountMinLen));
            return;
        }

        if (!RegexUtils.isPhone(account) && !RegexUtils.isEmail(account)) {
            showToast(R.string.account_format_err);
            return;
        }

        int pwdMinLen = getResources().getInteger(R.integer.password_min_len);
        String pwd = mPwdET.getText().toString().trim();
        String confirmPwd = mPwdConfirmET.getText().toString().trim();
        if (pwd.length() < pwdMinLen || confirmPwd.length() < pwdMinLen) {
            showToast(getString(R.string.pwd_len_limit, pwdMinLen));
            return;
        }

        if (!pwd.equals(confirmPwd)) {
            showToast(R.string.pwd_not_match);
            return;
        }

        mPresenter.register(account, pwd, confirmPwd);
    }

    @Override
    public void onRegisterSuccess(LoginBean loginBean) {
        showToast(R.string.register_success);
        finish();
    }

    @Override
    public void onRegisterFail(@ErrCode int code, Throwable throwable) {
        if (code == ErrCode.NET_ERR) {
            showToast(R.string.net_err);

        } else if (code == ErrCode.DATA_ERR) {
            showToast(R.string.account_pwd_err);

        } else if (code == ErrCode.SRC_ERR) {
            showToast(throwable.getMessage());

        } else {
            showToast(R.string.unknown_err);
        }
    }
}

File: app/src/main/java/com/ytempest/wanandroid/activity/register/RegisterActivity.kt
package com.ytempest.wanandroid.activity.register

import android.view.View
import com.ytempest.tool.util.RegexUtils
import com.ytempest.wanandroid.R
import com.ytempest.wanandroid.base.activity.MVVMActivity
import com.ytempest.wanandroid.base.createViewModel
import com.ytempest.wanandroid.base.vm.EntityObserver
import com.ytempest.wanandroid.databinding.ActivityRegisterBinding
import com.ytempest.wanandroid.http.ErrCode
import com.ytempest.wanandroid.http.bean.LoginBean
import com.ytempest.wanandroid.listener.TextWatcherListener
import com.ytempest.wanandroid.utils.CommonUtils
import com.ytempest.wanandroid.utils.SpaceInputFilter
import com.ytempest.wanandroid.utils.StatusBarUtil

/**
 * @author heqidu
 * @since 21-2-22
 */
class RegisterActivity : MVVMActivity<ActivityRegisterBinding>(), IRegisterView, View.OnClickListener {

    override val viewModel by lazy { createViewModel<RegisterViewModel>() }

    override fun onViewCreated() {
        StatusBarUtil.immersive(this)
        StatusBarUtil.setPaddingSmart(this, binding.root)

        initViews()
        initData()
    }

    private fun initViews() {
        val filters = arrayOf(SpaceInputFilter())
        val textWatcher = object : TextWatcherListener() {
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                super.onTextChanged(s, start, before, count)
                if (binding.accountEt.length() > 0 && binding.pwdEt.length() > 0 && binding.pwdConfirmEt.length() > 0) {
                    binding.registerBtn.setNormalStatus()
                } else {
                    binding.registerBtn.setDisableStatus()
                }
            }
        }

        arrayOf(binding.accountEt, binding.pwdEt, binding.pwdConfirmEt).forEach {
            it.filters = filters
            it.addTextChangedListener(textWatcher)
        }

        CommonUtils.showSoftKeyBoard(binding.accountEt)

        binding.backView.setOnClickListener { finish() }
        binding.registerBtn.setOnClickListener(this)
    }

    private fun initData() {
        viewModel.registerResult.observe(this, EntityObserver(
                onSuccess = { entity ->
                    onRegisterSuccess(entity.data)
                },
                onFail = { entity ->
                    onRegisterFail(entity.code, entity.throwable)
                },
                onComplete = {
                    stopLoading()
                }
        ))
    }

    override fun onRegisterSuccess(loginBean: LoginBean) {
        showToast(R.string.register_success)
        finish()
    }

    override fun onRegisterFail(@ErrCode code: Int, throwable: Throwable?) = when (code) {
        ErrCode.NET_ERR -> showToast(R.string.net_err)
        ErrCode.DATA_ERR -> showToast(R.string.account_pwd_err)
        ErrCode.SRC_ERR -> showToast(throwable?.message)
        else -> showToast(R.string.unknown_err)
    }

    override fun onClick(v: View) {
        when (v) {
            binding.registerBtn -> {
                onRegisterConfirmClick()
            }
        }
    }

    private fun onRegisterConfirmClick() {
        val accountMinLen = resources.getInteger(R.integer.account_min_len)
        val account: String = binding.accountEt.text.toString().trim()
        if (account.length < accountMinLen) {
            showToast(getString(R.string.account_len_limit, accountMinLen))
            return
        }

        if (!RegexUtils.isPhone(account) && !RegexUtils.isEmail(account)) {
            showToast(R.string.account_format_err)
            return
        }

        val pwdMinLen = resources.getInteger(R.integer.password_min_len)
        val pwd = binding.pwdEt.text.toString().trim()
        val confirmPwd: String = binding.pwdConfirmEt.text.toString().trim()
        if (pwd.length < pwdMinLen || confirmPwd.length < pwdMinLen) {
            showToast(getString(R.string.pwd_len_limit, pwdMinLen))
            return
        }

        if (pwd != confirmPwd) {
            showToast(R.string.pwd_not_match)
            return
        }

        showLoading()
        viewModel.register(account, pwd, confirmPwd)
    }

}